File: packages/shared/src/schemas.ts
import { z } from "zod";

// Mirror of packages/shared/src/fii_shared/schemas.py. Keep in sync by hand for now;
// we'll codegen from the Pydantic schemas in a later section if drift becomes a problem.

export const AgentRunStatus = z.enum([
  "pending",
  "running",
  "succeeded",
  "failed",
  "canceled",
]);
export type AgentRunStatus = z.infer<typeof AgentRunStatus>;

export const Ticker = z.object({
  symbol: z.string().min(1).max(10),
  exchange: z.string().max(16).nullable().optional(),
});
export type Ticker = z.infer<typeof Ticker>;

export const DeepDiveRequest = z.object({
  ticker: Ticker,
  horizon_months: z.number().int().min(1).max(240).default(36),
  notes: z.string().max(2000).nullable().optional(),
});
export type DeepDiveRequest = z.infer<typeof DeepDiveRequest>;

export const AnalysisRun = z.object({
  run_id: z.string(),
  ticker: Ticker,
  status: AgentRunStatus,
  started_at: z.string().datetime(),
  finished_at: z.string().datetime().nullable().optional(),
  step_functions_execution_arn: z.string().nullable().optional(),
  cost_usd: z.number().min(0).nullable().optional(),
});
export type AnalysisRun = z.infer<typeof AnalysisRun>;

File: packages/shared/src/fii_shared/schemas.py
"""Canonical domain schemas.

Kept deliberately thin at MVP — we'll grow this as Sections 2+ flesh out the data model.
Mirrored by the TypeScript zod schemas in packages/shared/src/schemas.ts.
"""

from datetime import datetime
from enum import StrEnum

from pydantic import BaseModel, Field


class AgentRunStatus(StrEnum):
    PENDING = "pending"
    RUNNING = "running"
    SUCCEEDED = "succeeded"
    FAILED = "failed"
    CANCELED = "canceled"


class Ticker(BaseModel):
    """Canonical identifier for an equity. Symbol is normalized to upper-case."""

    symbol: str = Field(min_length=1, max_length=10)
    exchange: str | None = Field(default=None, max_length=16)


class DeepDiveRequest(BaseModel):
    """Request to kick off a full specialist-swarm analysis for a single ticker."""

    ticker: Ticker
    horizon_months: int = Field(default=36, ge=1, le=240)
    notes: str | None = Field(default=None, max_length=2000)


class AnalysisRun(BaseModel):
    """Metadata for one deep-dive execution. Full reasoning trail lives in S3."""

    run_id: str
    ticker: Ticker
    status: AgentRunStatus
    started_at: datetime
    finished_at: datetime | None = None
    step_functions_execution_arn: str | None = None
    cost_usd: float | None = Field(default=None, ge=0)
